code/xcvm/cosmwasm/contracts/gateway/src/contract/ibc/ics20.rs

```diff
@@ -35,68 +39,137 @@ pub(crate) fn handle_bridge_forward(
+
+		let packet = XcPacket {
+			interpreter: String::from(info.sender).into_bytes(),
+			user_origin: msg.interpreter_origin.user_origin.clone(),
+			salt: msg.msg.salt,
+			program: msg.msg.program,
+			assets: vec![asset].into(),
+		};
+
+		deps.api.debug(&format!(
+			"cvm::gateway::ibc::ics20 route {}",
+			&serde_json_wasm::to_string(&route)?
+		));
+
+		let mut event = make_event("bridge")
+			.add_attribute("to_network_id", msg.to.to_string())
+			.add_attribute(
+				"assets",
+				serde_json_wasm::to_string(&packet.assets)
+					.map_err(|_| ContractError::FailedToSerialize)?,
+			)
+			.add_attribute(
+				"program",
+				serde_json_wasm::to_string(&packet.program)
+					.map_err(|_| ContractError::FailedToSerialize)?,
+			);
+
+		if !packet.salt.is_empty() {
+			let salt_attr = Binary::from(packet.salt.as_slice()).to_string();
+			event = event.add_attribute("salt", salt_attr);
+		}
 
-	let msg = to_cw_message(deps.as_ref(), deps.api, coin, route, packet, block)?;
+		let coin = Coin::new(amount.0, route.local_native_denom.clone());
+
+		match route.gateway_to_send_to.clone() {
+			GatewayId::CosmWasm { contract, .. } => {
+				let msg = to_cosmwasm_message(
+					deps.as_ref(),
+					deps.api,
+					coin,
+					route,
+					packet,
+					block,
+					contract,
+				)?;
+				(msg, event)
+			},
+			GatewayId::Evm { .. } => Err(ContractError::NotImplemented)?,
+		}
+	};
 
 	Ok(Response::default()
 		.add_event(event)
 		.add_submessage(SubMsg::reply_on_success(msg, ReplyId::TransportSent.into())))
 }
 
+/// When target network supports native cross chain operation of Transfer,
+/// and program as simple as just Transfer,
+/// can use instance of this structure to route pure funds transfer
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "std", derive(schemars::JsonSchema))]
+#[serde(rename_all = "snake_case")]
+pub struct IbcIcs20TransferShortcutRoute {
+	pub source: ChannelId,
+	pub denom: String,
+	pub sending: xc_core::transport::ibc::IbcIcs20Sender,
+}
+
+/// this method return route in case program can be just transfer
+pub fn ibc_ics_20_transfer_shortcut(
+	deps: Deps,
+	msg: &xc_core::gateway::BridgeForwardMsg,
+) -> Result<IbcIcs20TransferShortcutRoute, ContractError> {
+	let storage = deps.storage;
+	let this = load_this(storage)?;
+	let other = network::load_other(storage, msg.to)?;
+	let this_asset_id = msg.msg.assets.0[0].0;
+	let asset: AssetItem = state::assets::ASSETS
+		.load(storage, this_asset_id)
+		.map_err(|_| ContractError::AssetNotFoundById(this_asset_id))?;
+	if let Some(ics20) = other.connection.ics_20 {
+		if let Some(shortcut) = other.connection.use_shortcut {
+			if shortcut {
+				return Ok(IbcIcs20TransferShortcutRoute {
+					source: ics20.source,
+					denom: asset.local.denom(),
+					sending: this
+						.ibc
+						.expect("ibc")
+						.channels
+						.expect("channels")
+						.ics20
+						.expect("ics20")
+						.sender,
+				})
+			}
+		}
+	}
```

Comment:
It should be possible to flatten a bit:

```suggestion
	if let Some(ics20) = other.connection.ics_20 {
		if other.connection.use_shortcut == Some(true) {
			let sending = this
				.ibc
				.expect("ibc")
				.channels
				.expect("channels")
				.ics20
				.expect("ics20")
				.sender;
			return Ok(IbcIcs20TransferShortcutRoute {
				source: ics20.source,
				denom: asset.local.denom(),
				sending,
			})
		}
	}
```

Reply:
sure, will try in next PR.  it is possible to flatten even more tbh. 

just not sure what you typed compiles on version of Rust we use.